File: tic_tac_toe_verifier.test.ts
import {expect, test} from "bun:test";
import {prettyPrintGameReport, ticTacToeVerifier} from "./business_logic";
import type {CellValue} from "./domain";

// Edge case: n = 1
test("n=1, X wins", () => {
    const board = [["X"]] as CellValue[][];
    expect(prettyPrintGameReport(ticTacToeVerifier(board))).toBe("The board is valid. State: XWon. Winner: X.");
});

// O cannot win because X always goes first!
test("n=1, O wins", () => {
    const board = [["O"]] as CellValue[][];
    expect(prettyPrintGameReport(ticTacToeVerifier(board))).toBe(
        "The board is invalid: The number of 'X' tokens must be equal to or exactly one more than the number of 'O' tokens."
    );
});

test("n=1, empty", () => {
    const board = [["."]] as CellValue[][];
    expect(prettyPrintGameReport(ticTacToeVerifier(board))).toBe(
        "The board is valid. State: InProgress. Winner: No Winner."
    );
});

// Edge case: n = 2
test("n=2, X wins", () => {
    const board = [
        ["X", "X"],
        ["O", "."],
    ] as CellValue[][];
    expect(prettyPrintGameReport(ticTacToeVerifier(board))).toBe("The board is valid. State: XWon. Winner: X.");
});

// Invalid: implies O either "double dipped" on his turn or went first, both of which are illegal
test("n=2, O wins", () => {
    const board = [
        ["O", "X"],
        ["O", "."],
    ] as CellValue[][];
    expect(prettyPrintGameReport(ticTacToeVerifier(board))).toBe(
        "The board is invalid: The number of 'X' tokens must be equal to or exactly one more than the number of 'O' tokens."
    );
});

test("n=2, in progress", () => {
    const board = [
        ["X", "."],
        ["O", "."],
    ] as CellValue[][];
    expect(prettyPrintGameReport(ticTacToeVerifier(board))).toBe(
        "The board is valid. State: InProgress. Winner: No Winner."
    );
});

test("n=3, invalid: both players win", () => {
    const board = [
        ["X", "X", "X"],
        ["O", "O", "O"],
        ["X", "O", "X"],
    ] as CellValue[][];
    expect(prettyPrintGameReport(ticTacToeVerifier(board))).toBe(
        "The board is invalid: Both players cannot win simultaneously."
    );
});

test("n=3, X wins", () => {
    const board = [
        ["X", "O", "X"],
        ["O", "X", "O"],
        ["O", "X", "X"],
    ] as CellValue[][];
    expect(prettyPrintGameReport(ticTacToeVerifier(board))).toBe("The board is valid. State: XWon. Winner: X.");
});

test("n=3, O wins", () => {
    const board = [
        ["O", "X", "X"],
        ["O", "X", "X"],
        ["O", "O", "."],
    ] as CellValue[][];
    expect(prettyPrintGameReport(ticTacToeVerifier(board))).toBe("The board is valid. State: OWon. Winner: O.");
});

test("n=3, draw", () => {
    const board = [
        ["X", "O", "X"],
        ["O", "X", "X"],
        ["O", "X", "O"],
    ] as CellValue[][];
    expect(prettyPrintGameReport(ticTacToeVerifier(board))).toBe("The board is valid. State: Draw. Winner: No Winner.");
});

test("n=3, invalid: too many X's", () => {
    const board = [
        ["X", "X", "X"],
        ["O", "O", "X"],
        [".", ".", "."],
    ] as CellValue[][];
    expect(prettyPrintGameReport(ticTacToeVerifier(board))).toBe(
        "The board is invalid: The number of 'X' tokens must be equal to or exactly one more than the number of 'O' tokens."
    );
});

test("n=4, X wins with a row, valid token count", () => {
    const board = [
        ["X", "X", "X", "X"],
        ["O", "O", ".", "."],
        ["O", ".", ".", "."],
        [".", ".", ".", "."],
    ] as CellValue[][];
    expect(prettyPrintGameReport(ticTacToeVerifier(board))).toBe("The board is valid. State: XWon. Winner: X.");
});

test("n=4, O wins diagonal, valid token count", () => {
    const board = [
        ["O", ".", ".", "."],
        ["X", "O", ".", "."],
        ["X", "X", "O", "."],
        ["X", ".", ".", "O"],
    ] as CellValue[][];
    expect(prettyPrintGameReport(ticTacToeVerifier(board))).toBe("The board is valid. State: OWon. Winner: O.");
});

test("n=4, draw", () => {
    const board = [
        ["X", "X", "X", "O"],
        ["O", "O", "X", "O"],
        ["X", "O", "X", "O"],
        ["O", "X", "O", "X"],
    ] as CellValue[][];
    expect(prettyPrintGameReport(ticTacToeVerifier(board))).toBe("The board is valid. State: Draw. Winner: No Winner.");
});

test("illegal board: not square", () => {
    const board = [
        ["X", "O", "X"],
        ["O", "X"],
        ["O", "X", "X"],
    ] as unknown as CellValue[][];
    expect(prettyPrintGameReport(ticTacToeVerifier(board))).toBe(
        "The board is illegal: Input must be a non-empty square matrix"
    );
});

test("illegal board: empty", () => {
    const board = [] as unknown as CellValue[][];
    expect(prettyPrintGameReport(ticTacToeVerifier(board))).toBe(
        "The board is illegal: Input must be a non-empty square matrix"
    );
});

File: test/ticTacToeVerifierTest.fs
module TicTacToeVerifierTests

open Expecto
open TicTacToeVerifier.Domain
open TicTacToeVerifier.BusinessLogic

let tests =
    testList
        "TicTacToeVerifier tests"
        [ testList
              "Illegal board"
              [ testCase "Empty board"
                <| fun _ ->
                    let board = [||] |> Array.map (fun _ -> [||])

                    Expect.equal
                        (board |> ticTacToeVerifier |> prettyPrintGameReport)
                        "Illegal board. Board must be non-empty."
                        "Expected the board to be illegal due to being empty."

                testCase "Non-square board"
                <| fun _ ->
                    let board = [| [| Player X; Player O |]; [| Player O |] |]

                    Expect.equal
                        (board |> ticTacToeVerifier |> prettyPrintGameReport)
                        "Illegal board. Board must be square."
                        "Expected the board to be illegal due to not being square." ]
          testList
              "Edge cases"
              [ testCase "n=1, X wins"
                <| fun _ ->
                    let board = [| [| Player X |] |]

                    Expect.equal
                        (board |> ticTacToeVerifier |> prettyPrintGameReport)
                        "Valid board. Game State: Win for X."
                        "Expected the board to be valid with X winning."

                testCase "n=1, O wins"
                <| fun _ ->
                    let board = [| [| Player O |] |]

                    Expect.equal
                        (board |> ticTacToeVerifier |> prettyPrintGameReport)
                        "Invalid board. Invalid token counts: X has 0 tokens, O has 1 tokens. In valid games, X token count **must** be equal to or **exactly one** more than O token count."
                        "Expected the board to be invalid due to O winning without X."

                testCase "n=1, empty"
                <| fun _ ->
                    let board = [| [| Empty |] |]

                    Expect.equal
                        (board |> ticTacToeVerifier |> prettyPrintGameReport)
                        "Valid board. Game State: Ongoing."
                        "Expected the board to be valid and in progress with no winner."

                testCase "n=2, X Wins"
                <| fun _ ->
                    let board = [| [| Player X; Player X |]; [| Player O; Empty |] |]

                    Expect.equal
                        (board |> ticTacToeVerifier |> prettyPrintGameReport)
                        "Valid board. Game State: Win for X."
                        "Expected the board to be valid with X winning."

                // Invalid board: this means O either "double dipped" on his turn or went first, both of which are illegal
                testCase "n=2, O Wins"
                <| fun _ ->
                    let board = [| [| Player O; Player O |]; [| Player X; Empty |] |]

                    Expect.equal
                        (board |> ticTacToeVerifier |> prettyPrintGameReport)
                        "Invalid board. Invalid token counts: X has 1 tokens, O has 2 tokens. In valid games, X token count **must** be equal to or **exactly one** more than O token count."
                        "Expected the board to be invalid due to O winning without X."

                testCase "n=2, Ongoing"
                <| fun _ ->
                    let board = [| [| Player X; Empty |]; [| Player O; Empty |] |]

                    Expect.equal
                        (board |> ticTacToeVerifier |> prettyPrintGameReport)
                        "Valid board. Game State: Ongoing."
                        "Expected the board to be valid and in progress with no winner." ]

          testCase "n=3, invalid: both players win"
          <| fun _ ->
              let board =
                  [| [| Player X; Player X; Player X |]
                     [| Player O; Player O; Player O |]
                     [| Player X; Player O; Player X |] |]

              Expect.equal
                  (board |> ticTacToeVerifier |> prettyPrintGameReport)
                  "Invalid board. Both players cannot win simultaneously."
                  "Expected the board to be invalid due to both players winning simultaneously."

          testCase "n=3, invalid: too many X's"
          <| fun _ ->
              let board =
                  [| [| Player X; Player X; Player X |]
                     [| Player O; Player O; Player X |]
                     [| Empty; Empty; Empty |] |]

              Expect.equal
                  (board |> ticTacToeVerifier |> prettyPrintGameReport)
                  "Invalid board. Invalid token counts: X has 4 tokens, O has 2 tokens. In valid games, X token count **must** be equal to or **exactly one** more than O token count."
                  "Expected the board to be invalid due to too many X's."

          testCase "n=3, X wins"
          <| fun _ ->
              let board =
                  [| [| Player X; Player O; Player X |]
                     [| Player O; Player X; Player O |]
                     [| Player O; Player X; Player X |] |]

              Expect.equal
                  (board |> ticTacToeVerifier |> prettyPrintGameReport)
                  "Valid board. Game State: Win for X."
                  "Expected the board to be valid with X winning."

          testCase "n=3, O wins"
          <| fun _ ->
              let board =
                  [| [| Player O; Player X; Player X |]
                     [| Player O; Player X; Player X |]
                     [| Player O; Player O; Empty |] |]

              Expect.equal
                  (board |> ticTacToeVerifier |> prettyPrintGameReport)
                  "Valid board. Game State: Win for O."
                  "Expected the board to be valid with O winning."

          testCase "n=3, Draw"
          <| fun _ ->
              let board =
                  [| [| Player X; Player O; Player X |]
                     [| Player O; Player X; Player X |]
                     [| Player O; Player X; Player O |] |]

              Expect.equal
                  (board |> ticTacToeVerifier |> prettyPrintGameReport)
                  "Valid board. Game State: Draw."
                  "Expected the board to be valid with a draw."

          testCase "n=4, X wins"
          <| fun _ ->
              let board =
                  [| [| Player X; Player X; Player X; Player X |]
                     [| Player O; Player O; Empty; Empty |]
                     [| Player O; Empty; Empty; Empty |]
                     [| Empty; Empty; Empty; Empty |] |]

              Expect.equal
                  (board |> ticTacToeVerifier |> prettyPrintGameReport)
                  "Valid board. Game State: Win for X."
                  "Expected the board to be valid with X winning."

          testCase "n=4, O wins"
          <| fun _ ->
              let board =
                  [| [| Player O; Empty; Empty; Empty |]
                     [| Player X; Player O; Empty; Empty |]
                     [| Player X; Player X; Player O; Empty |]
                     [| Player X; Empty; Empty; Player O |] |]

              Expect.equal
                  (board |> ticTacToeVerifier |> prettyPrintGameReport)
                  "Valid board. Game State: Win for O."
                  "Expected the board to be valid with O winning."

          testCase "n=4, Draw"
          <| fun _ ->
              let board =
                  [| [| Player X; Player X; Player X; Player O |]
                     [| Player O; Player O; Player X; Player O |]
                     [| Player X; Player O; Player X; Player O |]
                     [| Player O; Player X; Player O; Player X |] |]

              Expect.equal
                  (board |> ticTacToeVerifier |> prettyPrintGameReport)
                  "Valid board. Game State: Draw."
                  "Expected the board to be valid and in progress with no winner." ]
